source/common/config/utility.h

```diff
@@ -117,6 +117,39 @@ class Utility {
   static void checkLocalInfo(const std::string& error_prefix,
                              const LocalInfo::LocalInfo& local_info);
 
+  /**
+   * Check the existence of a path for a filesystem subscription. Throws on error.
+   * @param path the path to validate.
+   */
+  static void checkFilesystemSubscriptionBackingPath(const std::string& path);
+
+  /**
+   * Check the validity of a cluster backing an api config source. Throws on error.
+   * @param clusters the clusters currently loaded in the cluster manager.
+   * @param api_config_source the config source to validate.
+   */
+  template <class ResourceType>
+  static void checkApiConfigSourceSubscriptionBackingCluster(
+      const Upstream::ClusterManager::ClusterInfoMap& clusters,
+      const envoy::api::v2::ApiConfigSource& api_config_source) {
+    if (api_config_source.cluster_name().size() != 1) {
+      // TODO(htuch): Add support for multiple clusters, #1170.
+      throw EnvoyException(
+          "envoy::api::v2::ConfigSource must have a singleton cluster name specified");
+    }
+
+    const auto& cluster_name = api_config_source.cluster_name()[0];
+    const auto& it = clusters.find(cluster_name);
+    if (it == clusters.end() || it->second.get().info()->addedViaApi() ||
+        (typeid(ResourceType) == typeid(envoy::api::v2::ClusterLoadAssignment) &&
```

Comment:
Sorry I just saw this. What is the reason for this check? IMO no resource type should depend on an EDS cluster even though it's technically possible. I would just make all xDS use fully static cluster types until if/when someone complains about it.

Reply:
Reasoning was that the most confusing dependency was EDS where we could even get in cyclical dependencies. But yes, I agree that to keep in line with the philosophy of the change (restrict all xDS deps equally) we could keep all xDS dependent on **fully** static clusters -- both non loaded via API and endpoints not done via EDS. I'll adjust.

 I'll also make a follow up doc PR to document the restriction.

Reply:
+1, thanks!